WIP tutorial documentation (#4)
pivotal-cf/docs-build-service: tutorial.md

---
title: Tutorial for Pivotal Build Service
owner: Partners
---

This topic describes how to get started with Pivotal Build Service.

> ### <a id='install'></a> Targeting and Logging into a running Pivotal Build Service

Reviewer: This seems a little too verbose. Maybe something like "**Configuring CLI**"

Once you have deployed Pivotal Build Service, the `pb` cli can be used to target it with the following commands:
```bash
pb api set <BUILD-SERVICE-API>
```
An example of the `<BUILD-SERVICE-API>` would be `https://build-service.example.com`. Use the `--skip-ssl-validation` flag if the Pivotal Build Service targets a UAA that has an self-signed CA cert.

You can confirm the targeted Pivotal Build Service using the below command:
```bash
pb api get
```

Once you are targeting the intended Pivotal Build Service, you can login to it as follows:

```bash
pb login
```

This command prompts you for a `username` and `password` which correspond to your UAA credentials. Additionally, the username and password can be passed in via environment variables with the following names:
`BUILD_SERVICE_USERNAME` and `BUILD_SERVICE_PASSWORD`. The cli will default to pick these up and use them if they exist in the environment.  

> ### <a id='install'></a> Creating a `team`

Reviewer: Anchors need to be adjusted. They all have `install`

Author: good catch

> A `team` is an entity on Pivotal Build Service that is used to managed authentication for the images built by Pivotal Build Service and to manage registry and git credentials for the images managed by said team.

Reviewer: typo: `managed` => `manage`

> A `team` is an entity on Pivotal Build Service that is used to managed authentication for the images built by Pivotal Build Service and to manage registry and git credentials for the images managed by said team.

Reviewer: It makes it seem like `team` is a PBS construct but if I understand this correctly user management is actually done via [UAA](https://docs.run.pivotal.io/concepts/architecture/uaa.html) so maybe more clarification, or explanation on the separation, of responsibilities between a `UAA team` and a `PBS team` is necessary.

Author: The UAA is the auth gateway. UAA and Build Service both have concepts of teams, both of which are sorta different. UAA has full blown support for teams and currently Build Service uses teams to provide some kind of isolation between images and users and ensure only images creators have access to the images they created. I don't think the concepts will map 1-1 between the 2 teams.

We can think of how we want to rephrase this to call that out though.

Only the users that belong to a team will be allowed to create images against said team. Additionally, they will be the only ones who can check the builds against an image.

You can configure a team using a file with the following `yaml` structure.

```yaml
name: example-team-name
registries:
- registry: registry.default.com
  username: <registry username>
  password: <registry password>
- registry: example.artifactory.com
  username: <artifactory username>
  password: <artifactory password>
repositories:
- domain: github.com
  username: <github-username>
  password: <password-for-github-user>
```

The registry credentials provided should belong to a user with `write` access on the registry. 

Save this file as `<example-team>.yaml` which can then be provided to Pivotal Build Service:

```bash
pb team apply -f /path/to/<example-team>.yaml
```

If the operation is successful, the cli will display the message: `Successfully applied team example-team-name`

> **Constraints:**

Reviewer 2: Maybe "Alpha Constraints" or "Current Constraints" instead, to convey that this will not always be this way?

1. Only one user per team

> 1. Cannot reference UAA user groups

Reviewer: Maybe this is simply my lack of domain knowledge but these constraints are a tad confusing to me.

1. "Only one user per team"
    * What user are we talking about here?
    * If there is only one user in a team, is it really a team? (only kind of joking)
2. "Cannot reference UAA user groups"
    * Are there plans on supporting this or else why would anyone expect for this to work?

Author: Yeah. At this point of time, a time is a concept that we need to expand on. 1 team <=> 1 user
It was to get us bootstrapped with an auth model. That is a very good question though. 

Given we expect UAA to be the source of user auth and the CF users will be familiar with the concept of user groups on UAA, this is to callout features we don't support that familiar CF users might expect.

Author: Any recommendations on how this should be reframed? Though I do realize it is more of features we don't have and not really the phrasing that is a problem here.

### <a id='install'></a> Creating an `image`

An image defines the specification that Pivotal Build Service uses to create images for a user. Here is an example of an image configuration: 

```yaml
team: example-team-name
source:
  git:
    url: https://github.com/example-org/sample-node-app
    revision: master
image:
  tag: registry.default.com/my-team-folder/node-app-image
```

> It is composed of the following components:

Reviewer 2: You might consider a format like we do for builder config TOMLs to lay out the purpose of each field:

https://buildpacks.io/docs/using-pack/working-with-builders/#builder-configuration

1. The `team` that the image belongs to. It has to be the team you are a part of as well. You can only create images for teams you belong to.
1. The `source` defines the git location of the code that images will be built against. The `revision` can either be a branch, tag or a commit-sha. When targeted against a branch, a build is triggered for every new commit. In case this is a private git repo, its credentials must be specified in the `repositories` section of the team configuration.

> 1. An `image registry` defines the destination registry of the builds for the image. The credentials for the target registry must be specified in the `registries` section of the team configuration. This should also match the domain of one of the registries provided in the team configuration.

Reviewer 2: Suggest "The `image.tag` key defines..." instead

The value of `image.tag` will be used to refer to the image once it has been created within Pivotal Build Service. Updating this field will lead to the creation of a new image.

The above image configuration can be saved as `<my-example-image>.yaml`

The configuration of the image can be applied to Pivotal Build Service:

```bash
pb image apply -f /path/to/<my-example-image>.yaml
```

> Pivotal Build Service auto-rebuilds images when one or more of the following things change:

Reviewer 2: Suggest ending this line with `when one or more of the following changes occurs:`

1. New buildpack versions are made available through an updated builder image

> 1. New commit on a branch or tag Pivotal Build Service is tracking

Reviewer 2: Maybe:

> A new commit appears on a branch or tag that Pivotal Build Service is tracking

to make it the same tense as the above bullet.

> 1. Updating the commit, branch, or git repo on the image's configuration file and re-applying it via `pb image apply`

Reviewer 2: Maybe:

> An update is made to the git repo details in the image's configuration file (and that change is applied via `pb image apply`)

> **Constraints:**

Reviewer 2: Same comment as above about wording here

> 1. Users can only specify source code that lives in a git repository

Reviewer 2: Are subpaths supported within the repo?

1. Pivotal Build Service does not rebuild images based on new OS packages (like cflinuxfs3)

### <a id='install'></a> Monitoring `builds` for an `image`

You can list all the builds created for an image with:

```bash
pb image builds <image-tag>
```

> The `<image-tag>` in the above command is the value of the field `image.tag` in the image's configuration. The output of the command might look similar to what's described below:

Reviewer 2: ~what's~ what is

Should avoid contractions :)

Alternately:

> ...might look similar to the output below:

```
Build    Status     Image       Started Time           Finished Time
-----    ------     -----       ------------           -------------
    1    SUCCESS    f5a1725f    2019-07-08 21:55:27    2019-07-08 21:56:54
    2    SUCCESS    428fe93e    2019-07-08 21:56:55    2019-07-08 21:57:40
    3    FAILED       --        2019-07-08 21:58:55    2019-07-08 21:59:40
    4    BUILDING     --        2019-07-08 21:58:55          --
    -    PENDING      --              --                     --
```

1. The `Build` column describes the index of builds in the order that they were built.
1. The `Status` column describes the status of a previous or a running/pending build.
1. The `Image` column contains the SHA256 of the image successfully built.

> 1. The  `Started Time` and `Finished Time` columns described when a build was kicked off and when it was completed.

Reviewer 2: ~described~ describe

[build-step-export] Reusing layer 'app' with SHA sha256:02e0070ce11bac1829174ec1296dcb1f3f04a4c30a958e2c41ad5498f78898fe
...
[build-step-export] Reusing layer 'org.cloudfoundry.springautoreconfiguration:auto-reconfiguration' with SHA sha256:93d94baf6d0dfc4981eb7d8ddfc4ae51f5c13cf87789b64ae8c4b015318a1b43
[build-step-export] *** Images:
[build-step-export]       registry.com/sample/demo:latest - succeeded
[build-step-export]       registry.com/sample/demo:b2.20190709.145448 - succeeded
[build-step-export]
[build-step-export] *** Digest: sha256:48a4ca8e4d8e8a9af26437588d0ce0e9d5c09b53aeb3ef64230a3d58d4b0dc90
[build-step-export]
[build-step-cache] Reusing layer 'io.pivotal.openjdk:openjdk-jdk' with SHA sha256:5554c7c06a266eb44a7cbdf0ecfaa14070e21af2b0bdfd1edd3b96f5168cd511
[build-step-cache] Reusing layer 'io.pivotal.buildsystem:build-system-cache' with SHA sha256:3b03fdd870a2dc1e924a040b604c25b76efafc1324ceb08eae8eae686fc3a940
...
[build-step-cache] Reusing layer 'org.cloudfoundry.springboot:spring-boot' with SHA sha256:effa8b80729cafa9f9a01b21a4badb5203510de0bb2e6b309ffd2593b0a28de7
[build-step-cache]
```

> This is the output of a successful build. Failed builds should indicate the reason of the failure in the logs and can be used during debugging.

Reviewer 2: Suggest removing:

> and can be used during debugging

The logs for a running build can be followed by using the `-f` flag, much like this:

```bash
pb image logs <image-tag> -b <build-number> -f
```
This should follow along with the progress of the build and terminate when the build completes.

> **Constraints:**

Reviewer 2: Same comment as above for wording here

> Pivotal Build Service stores the 10 most recent successful builds and 10 most recent failed builds.

Reviewer 2: Suggest:

> the ~10~ ten most recent successful builds and _the_ ~10~ ten most recent failed builds

### <a id='install'></a> Deleting `teams` and `images`

> The commands to delete a team and an image are similar. To delete an image run:

Reviewer 2: Could use some clarification: similar to what? Each other or their `apply` counterparts?

```bash
pb image delete <image-tag>
```

If the operation is successful, the cli will display the message: `Successfully deleted image <image-tag>`

> This will delete all the builds that belong to the Pivotal Build Service image. This **WILL NOT** delete the images that have been generated by those builds. To delete those images, one would have to do them manually from the registry.

Reviewer 3: ```suggestion
This will delete all the builds that belong to the Pivotal Build Service image. This **WILL NOT** delete the images that have been generated by those builds. To delete those images, one would have to delete each of them manually from the registry.
```

> This will delete all the builds that belong to the Pivotal Build Service image. This **WILL NOT** delete the images that have been generated by those builds. To delete those images, one would have to do them manually from the registry.

Reviewer 2: Suggest:

> ~do~ _delete_ them manually

---
title: Tutorial for Pivotal Build Service
owner: Partners
---

This topic describes how to get started with Pivotal Build Service.

### <a id='install'></a> Targeting and Logging into a running Pivotal Build Service

Once you have deployed Pivotal Build Service, the `pb` cli can be used to target it with the following commands:
```bash
pb api set <BUILD-SERVICE-API>
```
An example of the `<BUILD-SERVICE-API>` would be `https://build-service.example.com`. Use the `--skip-ssl-validation` flag if the Pivotal Build Service targets a UAA that has an self-signed CA cert.

You can confirm the targeted Pivotal Build Service using the below command:
```bash
pb api get
```

> Once you are targeting the intended Pivotal Build Service, you can login to it as follows:

Reviewer 2: ~login to~ log into it...

```bash
pb login
```

This command prompts you for a `username` and `password` which correspond to your UAA credentials. Additionally, the username and password can be passed in via environment variables with the following names:
`BUILD_SERVICE_USERNAME` and `BUILD_SERVICE_PASSWORD`. The cli will default to pick these up and use them if they exist in the environment.  

### <a id='install'></a> Creating a `team`

> A `team` is an entity on Pivotal Build Service that is used to managed authentication for the images built by Pivotal Build Service and to manage registry and git credentials for the images managed by said team.

Reviewer 2: used to ~managed~ manage...

**All the credentials required during image creation need to be a part of the team configuration. This includes registry credentials for the built images and repository credentials for the source code if it lies in a private repository**

> Only the users that belong to a team will be allowed to create images against said team. Additionally, they will be they only ones who can monitor the builds against an image.

Reviewer 4: ```suggestion
Only the users that belong to a team will be allowed to create images against said team. Additionally, they will be the only ones who can check the builds against an image.
```

You can configure a team using a file with the following `yaml` structure.

```yaml
name: example-team-name
registries:
- registry: registry.default.com
  username: <username with write access to above registry>

>   password: <password for above user>

Reviewer: These two template entries look a little awkward (`username with write access to above registry`, `password for above user`). I like the way the ones below for artifactory are provided. So something more along the lines of `registry username` and `registry password` would seem more consistent.

Also, the notion about the registry credentials needing write permissions should probably be mentioned outside of the sample/template.

- registry: example.artifactory.com
  username: <artifactory username>
  password: <artifactory user password>
repositories:
- domain: github.com
  username: <github-username>
  password: <password-for-github-user>
```

> Save this file as `<example-team>.yaml` which can then be provided to Build Service:

Reviewer 4: ```suggestion
Save this file as `<example-team>.yaml` which can then be provided to Pivotal Build Service:
```

```bash
pb team apply -f /path/to/<example-team>.yaml
```

If the operation is successful, the cli will display the message: `Successfully applied team example-team-name`

**Constraints:**
1. Only one user per team
1. Cannot reference UAA user groups

### <a id='install'></a> Creating an `image`

> An image defines the specification that Build Service uses to create images for a user. Here is an example of an image configuration: 

Reviewer 4: ```suggestion
An image defines the specification that Pivotal Build Service uses to create images for a user. Here is an example of an image configuration: 
```

```yaml
team: example-team-name
source:
  git:
    url: https://github.com/example-org/sample-node-app
    revision: master
image:
  tag: registry.default.svc.cluster.io/my-team-folder/node-app-image
```

It is composed of the following components:

1. The `team` that the image belongs to. It has to be the team you are a part of as well. You can only create images for teams you belong to.

> 1. The `source` that defines that src that images will be built against. The revision can either be a branch or a commit-sha. When targeted against a branch, a build is triggered for every new commit. In case this is a private git repo, its credentials must be specified in the `repositories` section of the team configuration.

Reviewer 4: ```suggestion
1. The `source` that defines location where the code for that the image will be built against. The revision can either be a branch or a commit-sha. When targeted against a branch, a build is triggered for every new commit. In case this is a private git repo, its credentials must be specified in the `repositories` section of the team configuration.
```

1. An `image registry` that defines the destination registry of the builds for the image. The credentials for the target registry must be specified in the `registries` section of the team configuration.

The value of `image.tag` will be used to refer to the image once it has been created within Build Service. Updating this field will lead to the creation of a new image.

The above image configuration can be saved as `<my-example-image>.yaml`

> The configuration of the team can be applied to build service:

Reviewer 4: ```suggestion
The configuration of the image can be applied to Pivotal Build Service:
```

```bash
pb image apply -f /path/to/<my-example-image>.yaml
```

New builds of an images when one or more of the following things change:
1. New buildpack versions are made available through an updated builder image
1. New commit on a branch build service is tracking
1. Updating the commit, branch, or git repo on the image's configuration file and re-applying it via `pb image apply`

**Constraints:**

> 1. Users can only specify source code that lives in a git repo  

Reviewer 4: ```suggestion
1. Users can only specify source code that lives in a git repository
```

> 1. The Build Service Alpha does not rebuild images based on new OS packages (like cflinuxfs3)

Reviewer 4: ```suggestion
1. Pivotal Build Service Alpha does not rebuild images based on new OS packages (like cflinuxfs3)
```

> ### <a id='install'></a> Monitoring `builds` against an `image`

Reviewer 4: ```suggestion
### <a id='install'></a> Monitoring `builds` for an `image`
```

> You can list all the builds created against an image with:

Reviewer 4: ```suggestion
You can list all the builds created for an image with:
```

> 1. Pivotal Build Service Alpha does not rebuild images based on new OS packages (like cflinuxfs3)

Reviewer: Do readers of this document know what `Alpha` means? Would a better alternative be "The current version of Pivotal Build Service does not rebuild images based on new OS packages"? It can later be removed when it's supported.

Author: Yup. Dropped the alpha. And to your point, I would assume caveating it would imply we eventually would fix it?

Does `1. Pivotal Build Service does not rebuild images based on new OS packages (like cflinuxfs3)` work?

```
Build    Status     Image       Started Time           Finished Time
-----    ------     -----       ------------           -------------
    1    SUCCESS    f5a1725f    2019-07-08 21:55:27    2019-07-08 21:56:54
    2    SUCCESS    428fe93e    2019-07-08 21:56:55    2019-07-08 21:57:40
    3    FAILED       --        2019-07-08 21:58:55    2019-07-08 21:59:40
    4    BUILDING     --        2019-07-08 21:58:55          --
    -    PENDING      --              --                     --
```

1. The `Build` column describes the index of builds in the order that they were built.
1. The `Status` column describes the status of a previous or a running/pending build.
1. The `Image` column contains the SHA256 of the image successfully built.
1. The  `Started Time` and `Finished Time` columns described when a build was kicked off and when it was completed.

To get the logs of a particular build, run the following:

```bash
pb image logs <image-tag> -b <build-number>
```

The output of the command will look similar to this:

```

> [build-step-credential-initializer] {"level":"info","ts":1562684107.3441668,"logger":"fallback-logger","caller":"creds-init/main.go:40","msg":"Credentials initialized.","commit":"002a41a"}

Reviewer: Do we really need a full output sample? Maybe an abridge version would be sufficient. Similar lines can be replaced with an ellipsis `...`

eg.
```text
[build-step-detect] ======== Results ========
[build-step-detect] skip: Cloud Foundry Archive Expanding Buildpack
[build-step-detect] pass: Pivotal OpenJDK Buildpack
...
[build-step-detect] skip: Cloud Foundry JMX Buildpack
[build-step-detect] pass: Cloud Foundry Spring Auto-reconfiguration Buildpack
```

Author: Can a man not be lazy Javier ;)

That is valuable input

```bash
pb api get
```

Once you are targeting the intended Build Service, you can login to it as follows:

```bash
pb login
```

This command prompts you for a `username` and `password` which correspond to your UAA credentials. Additionally, the username and password can be passed in via env variable with the following names:
`BUILD_SERVICE_USERNAME` and `BUILD_SERVICE_PASSWORD`. The cli will default to pick these up and use them if they exist in the env.  

### <a id='install'></a> Creating a `team`

A `team` is an entity on Build Service that is used to managed auth for the images Build Service builds and to manage registry and git credentials for the images managed by said team.

**All the credentials required during image creation need to be a part of the team configuration. This includes registry credentials for the built images and repository credentials for the source code if it lies in a private repository**

Only the users that belong to a team will be allowed to create images against said team. Additionally, they will be they only ones who can monitor the builds against an image.

You can configure a team using a file with the following `yaml` structure.

> ```yaml

Reviewer 5: This is great. Would it be useful to also have schema level documentation such as https://concourse-ci.org/tasks.html?

name: example-team-name
registries:
- registry: registry.default.svc.cluster.io
  username: <username with write access to above registry>
  password: <password for above user>
- registry: example.artifactory.com
  username: <artifactory username>
  password: <artifactory user password>
repositories:
- domain: github.com
  username: <github-username>
  password: <password-for-github-user>
```

Save this file as `<example-team>.yaml` which can then be provided to Build Service:

```bash
pb team apply -f /path/to/<example-team>.yaml
```

If the operation is successful, the cli will display the message: `Successfully applied team example-team-name`

**Constraints:**
1. Only one user per team
1. Cannot reference UAA user groups

> ### <a id='install'></a> Creating an `image`

Reviewer 5: Is there a place there is documentation about what types of images the we can even build at this point?

Aka: Node, Python, Java. Basic documentation for each?

Reviewer 4: Our builder only support Node, Yarn and Java 11

An image defines the specification that Build Service uses to create images for a user. Here is an example of an image configuration: 

```yaml
team: example-team-name
source:
  git:

>     url: https://github.com/example-org/sample-node-app

Reviewer 5: Should we have a real sample app somewhere?

    revision: master
image:

>   tag: registry.default.svc.cluster.io/my-team-folder/node-app-image

Reviewer 5: Schema level definition of what an image tag is would be great. Especially how it relates to registry creds. 

`registry.default.svc.cluster.io` looks a bit like an kubernetes service? Are we unintentionally implying something here?

```

It is composed of the following components:

1. The `team` that the image belongs to. It has to be the team you are a part of as well. You can only create images for teams you belong to.
1. The `source` that defines that src that images will be built against. The revision can either be a branch or a commit-sha. When targeted against a branch, a build is triggered for every new commit. In case this is a private git repo, its credentials must be specified in the `repositories` section of the team configuration.
1. An `image registry` that defines the destination registry of the builds for the image. The credentials for the target registry must be specified in the `registries` section of the team configuration.

The value of `image.tag` will be used to refer to the image once it has been created within Build Service. Updating this field will lead to the creation of a new image.

The above image configuration can be saved as `<my-example-image>.yaml`

The configuration of the team can be applied to build service:

```bash
pb image apply -f /path/to/<my-example-image>.yaml
```

> New builds of an images when one or more of the following things change:

Reviewer 6: Slight grammar nit here, could be rewritten as "Build service executes automated rebuilds when one or more of the following things change:"

1. New buildpack versions are made available through an updated builder image

> 1. New commit on a branch build service is tracking

Reviewer 5: Branch or tag?

[build-step-cache] Caching layer 'org.cloudfoundry.buildpacks.npm:cache' with SHA sha256:01aa7d0778ae3900f69ded6a8e0b3b27c906fdf357a9eeef0b6abd052659942c
[build-step-cache] Caching layer 'org.cloudfoundry.buildpacks.npm:node_modules' with SHA sha256:f91092a7723e6a573564c673d32aea089deb1c3bd547c2f2888ed3f90853ed23
[build-step-cache]
```

This is the output of a successful build. Failed builds should indicate the reason of the failure in the logs and can be used during debugging.

The logs for a running build can be followed by using the `-f` flag, much like this:

```bash
pb image logs <image-tag> -b <build-number> -f
```
This should follow along with the progress of the build and terminate when the build completes.

**Constraints:**

Build Service stores the 10 most recent successful builds and 10 most recent failed builds.

### <a id='install'></a> Deleting `teams` and `images`

The commands to delete a team and an image are similar. To delete an image run:

```bash
pb image delete <image-tag>
```

If the operation is successful, the cli will display the message: `Successfully delete image <image-tag>`

> This will delete all the builds that belong. This **WILL NOT** delete the images that have been created by those builds. Presently, do delete those images, one would have to do them manually from the registry.

Reviewer 6: "Presently, to delete these images, one would have to remove them manually from the registry"

Similarly, team deletion can be performed using:

```bash
pb team delete <team-name>
```

> If the operation is successful, the cli will display the message: `Successfully applied team <team-name>`

Reviewer 6: Successfully applied team name?  Shouldn't it say "deleted"?

Author: yes, yes it should
